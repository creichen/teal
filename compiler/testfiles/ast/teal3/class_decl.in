class MyClass = {}

class MyClass [MyType1, MyType2] = {}

class MyClass (arg1, arg2) = {}

class MyClass <: Super1, Super2 = {}

class MyClass [MyType1, MyType2] (arg1, args2) = {}

class MyClass [MyType1, MyType2] <: Super1, Super2 [MyType3, MyType4] = {}

class MyClass (arg1, arg2) <: Super1, Super2 [MyType3, MyType4] = {}

class MyClass = {
      // variables
      var v : MyType;
      var w : T := v;

      // methods
      fun f(a) = return a;
}