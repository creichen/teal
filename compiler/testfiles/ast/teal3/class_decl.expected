Module
  List
  List
    ClassDecl
      IdDecl Identifier="MyClass"
      List
      List
      List
      List
    ClassDecl
      IdDecl Identifier="MyClass"
      List
        TypeVarDecl
          IdDecl Identifier="MyType1"
        TypeVarDecl
          IdDecl Identifier="MyType2"
      List
      List
      List
    ClassDecl
      IdDecl Identifier="MyClass"
      List
      List
      List
      List
        VarDecl
          IdDecl Identifier="arg1"
          Opt
          Opt
        VarDecl
          IdDecl Identifier="arg2"
          Opt
          Opt
    ClassDecl
      IdDecl Identifier="MyClass"
      List
      List
        Type
          IdUse Identifier="Super1"
          List
          List
        Type
          IdUse Identifier="Super2"
          List
          List
      List
      List
    ClassDecl
      IdDecl Identifier="MyClass"
      List
        TypeVarDecl
          IdDecl Identifier="MyType1"
        TypeVarDecl
          IdDecl Identifier="MyType2"
      List
      List
      List
        VarDecl
          IdDecl Identifier="arg1"
          Opt
          Opt
        VarDecl
          IdDecl Identifier="args2"
          Opt
          Opt
    ClassDecl
      IdDecl Identifier="MyClass"
      List
        TypeVarDecl
          IdDecl Identifier="MyType1"
        TypeVarDecl
          IdDecl Identifier="MyType2"
      List
        Type
          IdUse Identifier="Super1"
          List
          List
        Type
          IdUse Identifier="Super2"
          List
            Type
              IdUse Identifier="MyType3"
              List
              List
            Type
              IdUse Identifier="MyType4"
              List
              List
          List
      List
      List
    ClassDecl
      IdDecl Identifier="MyClass"
      List
      List
        Type
          IdUse Identifier="Super1"
          List
          List
        Type
          IdUse Identifier="Super2"
          List
            Type
              IdUse Identifier="MyType3"
              List
              List
            Type
              IdUse Identifier="MyType4"
              List
              List
          List
      List
      List
        VarDecl
          IdDecl Identifier="arg1"
          Opt
          Opt
        VarDecl
          IdDecl Identifier="arg2"
          Opt
          Opt
    ClassDecl
      IdDecl Identifier="MyClass"
      List
      List
      List
        VarDeclElt
          VarDecl
            IdDecl Identifier="v"
            Opt
              Type
                IdUse Identifier="MyType"
                List
                List
            Opt
        VarDeclElt
          VarDecl
            IdDecl Identifier="w"
            Opt
              Type
                IdUse Identifier="T"
                List
                List
            Opt
              Access
                IdUse Identifier="v"
        FunDeclElt
          FunDecl
            IdDecl Identifier="f"
            Opt
            List
              VarDecl
                IdDecl Identifier="a"
                Opt
                Opt
            Opt
              ReturnStmt
                Access
                  IdUse Identifier="a"
      List
